proposals/TP1-xds-transport-next.md

```diff
@@ -0,0 +1,1028 @@
+federation and removes special case treatment of aliasing from the xDS transport
+protocol.
+
+### Related Proposals:
+
+* The original Google Docs design documents:
+  [Full](https://docs.google.com/document/d/1zZav-IYxMO0mP2A7y5XHBa9v0eXyirw1CxrLLSUqWR8/edit#heading=h.w8hw2zbv3jwl)
+  [Summary](https://docs.google.com/document/d/1m5_Q9LUlzvDdImP0jqh1lSTrLMldsApTQX6ibbd3i7c/edit?ts=5ec2fd8a#heading=h.w8hw2zbv3jwl)
+* [xDS transport protocol.](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol)
+
+## Proposal
+
+The proposal below is fully backwards compatible with the existing xDS transport
+protocol and does not require any changes to clients or servers that do not wish
+to opt into the new naming scheme.
+
+### URI-based xDS resource names
+
+The following URI schema is proposed for xDS resource names:
+
+`xdstp://[{authority}]/{resource type}/{id/*}?{context parameters}{#processing directive,*}`
+
+* `authority` is an opaque string naming a resource authority, e.g.
+  `some.control.plane`.
+* `resource type` is the xDS resource type URL, e.g. `envoy.config.route.v3.RouteConfiguration`
+* `id/*` is the remaining path component of the URI and is fully opaque; naming
+  is at the discretion of the control plane(s).
+* `context parameters` are the URI query parameters and express contextual
+  information for selecting resource variants, for example `shard_id=1234&direction=inbound`.
+* A `processing directive` provides additional information to the client on how
+  the URI is to be interpreted. Two supported directives are:
+  * `alt=<xdstp:// URI>` expressing a fallback URI.
+  * `entry=<resource name>` providing an anchor referring to an inlined resource
+    for [list collections](#list). Resource names must be of the form `[a-zA-Z0-9_-\./]+`.
+
+Multiple directives are comma separated in the fragment component of the URI.
+
+An example application of the above URI scheme is:
+
+`xdstp://some.control.plane/envoy.config.route.v3.RouteConfiguration/foo/bar?shard_id=1234&direction=inbound`
+
+We distinguish between two forms of `xdstp://` URIs:
+
+* Uniform Resource Locators (URLs) instructing a client on how to locate a given
+  xDS resource. URLs may include processing directives.
+
+* Uniform Resource Names (URNs) referring to a specific xDS resource. URNs do not
+  include processing directives. `xdstp://` URNs provide self-contained cache keys.
+
+#### Proto3 expression
+
+`xdstp://` URIs are canonically encoded in `proto3` messages. The structure
+for this encoding is provided at https://github.com/cncf/xds/tree/main/xds/core/v3:
+
+* [URL](https://github.com/cncf/xds/blob/main/xds/core/v3/resource_locator.proto) definition
+* [URN](https://github.com/cncf/xds/blob/main/xds/core/v3/resource_name.proto) definition
+
+#### Normalization
+
+Two `xdstp://` URNs are considered equivalent if they match component-wise
+modulo context parameter ordering.
+
+## Context parameters
```

Comment:
How are we going to handle duplicate context params from different process? Say one from URL has same key with the one from dynamic context param added by client.

Reply:
@markdroth did we end up with any rules around how URL parameters (i.e. free form) are supposed to work? I.e. "they can never starts with xds." or "they can optionally start with xds. if deliberately overriding other context parameters"?

Reply:
I think it will be clear to have some rules for context parameters from URL.

Reply:
Sure. @markdroth can you comment on https://github.com/cncf/xds/pull/6#discussion_r611273789?

Reply:
I think where we wound up is that the ad-hoc context params can't start with `xds.`, so there will basically never be any conflicts.